### ISSUE TITLE
Rework package structure and documentation - [merged]

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#307]
_Merges use-api-autodoc -> v2_

#### The overall objectives

This MR proposes to modernize the package structure and documentation:

- Use an `src` layout for the package instead of a flat layout
- Update the `poetry.lock` file with Poetry v2, which will be required to install Leaspy
- Use the `auto-api` extension to generate automatically the API documentation
- Remove all the RST files which were basically doing what `auto-api` is doing for free
- Use the `myst_nb` extension to rely on Markdown rather than RST for writing documentation (it will be easier for contributors to write docs)
- Rewrite the existing documentation pages from RST to Markdown (I need to check that I didn't do too many mistakes doing that...)
- Run the formatting tools on all python files (it was the opportunity since the were all moved to the `src` folder...)

#### The package architectural changes

When moving to the `src` layout I faced a lot of circular import issues which are linked to the architecture of the package. Currently, these problems are masked by having some imports under an `if TYPE_CHECKING` block. In this MR, I removed all of these and had to restructure things a bit:

- The `AlgorithmSettings` are now inside of the `algo` module and not in `io.settings`
- The `ModelSettings` are now inside of the `models` module and not in `io.settings` (the latter was thus deleted)
- The `noise_models` module was deleted as well as related tests which is why we went from 9 failing tests to 4
- The public API of each module is declared in `__all__` and imports have been made more straightforward
- The `Loader` class in the module `datasets` have been replaced by simple functions
- The algo and model class factories were replaced by functions with enumerations (a much simpler pattern that we use a lot in other projects, we can of course discuss this point if you disagree)

#### The public "API"

What is currently called the "API" (i.e. the `Leaspy` class as well as the quick imports) was modified a bit. In my opinion the Leaspy class is useless and misleading and I really think it should disappear at some point. In this MR I kept it (just do `from leapsy.api import Leaspy` instead of `from leaspy import Leaspy`). The quick imports (like `from leapsy import Data` or `from leapsy import AlgorithmSettings` have been deleted and users should import from the right module: `from leapsy.io.data import Data` or `from leaps.algo import AlgorithmSettings`).

#### Next steps

There is still a lot of work to make the package structure better (the `io` module for example could be refactored), but since this is already a huge MR, I think we can target those in follow-up MRs.

### Where should the reviewer start ?

Honestly, the diff is almost impossible to make sense of. I think the easiest way is to get the branch locally and look at the code and structure in an IDE. The documentation changes can be analyzed by looking at the deployed doc and the new Markdown files.

### How can the code be tested ?

The tests are passing, except 4 tests that are already broken on V2 due to the Ordinal models. We went from 9 broken tests to 4 because I removed the noise models and the related tests.

### What issues are linked to the MR ?

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>df8644b5 - rework algo module</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/185/diffs?diff_id=1246726677&start_sha=4c876543620b7cbbe8ce22da473a08407ccca006)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>864acdbe - refactor</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/185/diffs?diff_id=1248147464&start_sha=df8644b5642cc3d4e307342ad6ed89630ee712b5)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>82a5466f - fix things</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/185/diffs?diff_id=1248676664&start_sha=864acdbead12c7b6e09f7bd5db67dc16eef01fa6)

**Comment by NicolasGensollen** [comment]
marked this merge request as **ready**

**Comment by NicolasGensollen** [comment]
requested review from @caglayantuna

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 17:45

Commented on [docs/glossary.md line 23](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..2c31e4237effc6afa6c6b60ef546120ab2bd4e8d#diff-64cb45c2fa3fa9eb495bd0a75f2392e2R23)



maybe we can change it to `fit` ?

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 17:45

Commented on [docs/nutshell.md line 8](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..82a5466fee29f74083e5f9c028479e613fb53b0f#diff-aca0d81889f742499237d28383bd956eR8)



this import won't be correct after this MR. Should we change now?

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 17:45

Commented on [docs/nutshell.md line 102](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..ee617b3df0504d127dd132a16ebe906544895406#diff-aca0d81889f742499237d28383bd956eR102)



This link is not correct. Maybe we can change now

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 17:45

Commented on [src/leaspy/models/factory.py line 80](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..2c31e4237effc6afa6c6b60ef546120ab2bd4e8d#diff-a0e5f253f8fd39bf2ffc495f2c17cf5fR80)



we had an error here before when user inputs a wrong name;

```python
            raise LeaspyModelInputError(
                "The name of the model you are trying to create does not exist! "
                f"It should be in {{{repr(tuple(ALL_MODELS.keys()))[1:-1]}}}"
```

maybe we can add an error in the end. WDYT?

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 17:45

Thanks @NicolasGensollen for this MR and detailed explanation. I did first a first review with very few questions. I have also some other general questions to ask from here. 

1 - I agree about misleading of Leaspy class name but also api name. Is it more useful to have leaspy.api? What do you think about the remove both of them and having leaspy as;
 ```python
import leaspy as lp
model = lp.model(...)
model.fit(....)
```
Is it too complicated to do? WDYT?

2- What about removing everything related to realizations in this MR. It still exists in io directory. As you said we can continue to edit io folder in the future but removing something could be easy. Because I guess that they are not used anymore in v2 

3- There are a lot of changes in this MR particularly imports and directory. However, we have also ongoing MRs in leaspy. Can we solve the conflicts easily if we merge this MR? 

4- I see that you removed settings and used them in algo and model. But where is outputssettings? Actually, it is used for fitoutputmanager. 

5- I couldn't understand how API reference is created with doc. Maybe we can talk about it later. As one concern for today, there are package contents inside of API reference which seems weird.

**Comment by NicolasGensollen** [comment]
I think the idea was to explain what the term calibration means (which, I agree, is basically the same as fit), but still I'd leave the entry in case users stumble across this word and need a definition.

**Comment by NicolasGensollen** [comment]
Ah yes indeed, good catch ! I'll change that !

**Comment by NicolasGensollen** [comment]
I'm OK to change it now, what URL do you think we should put ?

**Comment by NicolasGensollen** [comment]
With this pattern, assuming you handled all the enumeration variants, you don't need an `else` or an error raising at the end because there is no way you can get there. When casting the `name` variable from a string to a `ModelName`, if the `name` is not one of the allowed variants, then you'll get an error saying that "xxx is not a valid ModelName".

**Comment by NicolasGensollen** [comment]
1. I completely agree that the `Leaspy` class should disappear, and for A LOT of reasons. First, it is clearly misleading (never seen a package where you instantiate a class with the same name...). Second, it is useless because it is basically a wrapper around the model class (i.e. there is no real internal state that the class is modeling, and you can do the exact same thing with a model). Third, it creates a lot of circular import issues because objects buried deep in modules actually import this class to do stuff... I thought about deleting it in this MR, but I think this would trigger a lot of changes in the code (especially the tests), so I decided to postpone this move in a follow-up PR where it'd be easier to see the consequences. I simply removed the possibility to do `from leaspy import Leaspy` because it was hurting my eyes.

2. More or less the same reason, I'd prefer to do it in a dedicated MR where the consequences should be easy to spot and reason about. Theoretically, it should only be a matter of removing the folder and associated tests (just like for noise models).

3. Good point. This changes the structure a lot and currently opened MR will be difficult to update. Besides your MR on the FOM, I don't think we have a lot of complex MR that are likely to get merged one day (the simple ones could easily be re-created from the new structure). I will check the opened MR to be sure.

4. `OutputSettings` is in `Leaspy.algo.settings` for now because it seemed to be used only by `AlgorithmSettings`. We can continue to improve on this in follow-up PRs. For example, I think that the `FitOutputManager` would better belong within the algo module. 

5. Not sure to understand what you mean, I'll take a look at the rendered doc...

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 10:37

Commented on [docs/glossary.md line 23](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..2c31e4237effc6afa6c6b60ef546120ab2bd4e8d#diff-64cb45c2fa3fa9eb495bd0a75f2392e2R23)



Ok I understand, thanks

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 10:41

Commented on [docs/nutshell.md line 102](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..ee617b3df0504d127dd132a16ebe906544895406#diff-aca0d81889f742499237d28383bd956eR102)



Here, there is a notebook: https://gitlab.com/icm-institute/aramislab/leaspy/-/blob/master/example/start/starter.ipynb

**Comment by NicolasGensollen** [comment]
changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/185/diffs?diff_id=1250849577&start_sha=82a5466fee29f74083e5f9c028479e613fb53b0f#6df5cad27a5bd8e1d134aaef04edb3652868a3aa_8_8)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>ee617b3d - update docs/nutshell as if the Leaspy class didn&#39;t exist</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/185/diffs?diff_id=1250849577&start_sha=82a5466fee29f74083e5f9c028479e613fb53b0f)

**Comment by NicolasGensollen** [comment]
I updated the code samples from this page. I made as if the `Leaspy` class didn't exist (since I believe this is our objective). The API isn't perfect yet but that's a start. Let me know what you think.

**Comment by NicolasGensollen** [comment]
changed this line in [version 6 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/185/diffs?diff_id=1250855501&start_sha=ee617b3df0504d127dd132a16ebe906544895406#6df5cad27a5bd8e1d134aaef04edb3652868a3aa_102_102)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>2c31e423 - update notebook link in docs/nutshell</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/185/diffs?diff_id=1250855501&start_sha=ee617b3df0504d127dd132a16ebe906544895406)

**Comment by NicolasGensollen** [comment]
Made the update, thanks !

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 10:55

Commented on [docs/nutshell.md line 8](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..82a5466fee29f74083e5f9c028479e613fb53b0f#diff-aca0d81889f742499237d28383bd956eR8)



yes, it seems much better, thanks.

**Comment by NicolasGensollen** [comment]
I think the weird things at the end of the autogenerated API documentation are related to the stuffs we currently have in the base `__init__.py` file, i.e. it is not a problem with auto-api. I'll look at that later since it's not a major issue in my opinion.
I'm closing this thread, feel free to re-open if you think some points need more discussion.

**Comment by NicolasGensollen** [comment]
resolved all threads

**Comment by NicolasGensollen** [comment]
@caglayantuna If this looks good to you, please approve and I'll merge it so we can iterate.

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Feb 5, 2025, 10:33

no need to re-open, I understood your point

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Feb 5, 2025, 10:34

yes @NicolasGensollen , it is ok, we can merge this

**Comment by NicolasGensollen** [comment]
mentioned in commit 91e00577851de2eb625f05f18cd94256beca5fec
